cmd/podman/completion/completion.go

```diff
@@ -0,0 +1,368 @@
+	}
+)
+
+func init() {
+	registry.Commands = append(registry.Commands, registry.CliCommand{
+		Mode:    []entities.EngineMode{entities.ABIMode, entities.TunnelMode},
+		Command: completionCmd,
+	})
+}
+
+//Get containers with status.
+func getContainers(status string) ([]string, cobra.ShellCompDirective) {
+	var suggestions []string
+
+	listOpts := entities.ContainerListOptions{
+		Filters: make(map[string][]string),
+	}
+	listOpts.All = true
+
+	if status != "all" {
+		listOpts.Filters = map[string][]string{"status": {status}}
+	}
+
+	containers, err := registry.ContainerEngine().ContainerList(registry.GetContext(), listOpts)
+	if err != nil {
+		return nil, cobra.ShellCompDirectiveError
+	}
+
+	for _, container := range containers {
+		//include name in suggestions
+		//if ids should be also completed
+		//suggestions = append(suggestions, container.Names[0], container.ID[0:12])
+		suggestions = append(suggestions, container.Names...)
+	}
+
+	return suggestions, cobra.ShellCompDirectiveNoFileComp
+}
+
+//Get pods with status.
+func getPods(status string) ([]string, cobra.ShellCompDirective) {
+	var suggestions []string
+
+	listOpts := entities.PodPSOptions{
+		Filters: make(map[string][]string),
+	}
+
+	if status != "all" {
+		listOpts.Filters = map[string][]string{"status": {status}}
+	}
+
+	pods, err := registry.ContainerEngine().PodPs(registry.GetContext(), listOpts)
+	if err != nil {
+		return nil, cobra.ShellCompDirectiveError
+	}
+
+	for _, pod := range pods {
+		//include name in suggestions
+		//if ids should be also completed
+		//suggestions = append(suggestions, pod.Name, pod.Id[0:12])
+		suggestions = append(suggestions, pod.Name)
+	}
+
+	return suggestions, cobra.ShellCompDirectiveNoFileComp
+}
+
+//Get all volumes.
+func getVolumes() ([]string, cobra.ShellCompDirective) {
+	var suggestions []string
+
+	lsOpts := entities.VolumeListOptions{}
+
+	volumes, err := registry.ContainerEngine().VolumeList(context.Background(), lsOpts)
+	if err != nil {
+		return nil, cobra.ShellCompDirectiveError
+	}
+
+	for _, volume := range volumes {
+		suggestions = append(suggestions, volume.Name)
+	}
+
+	return suggestions, cobra.ShellCompDirectiveNoFileComp
+}
+
+//Get all images.
+func getImages(toComplete string) ([]string, cobra.ShellCompDirective) {
+	var suggestions []string
+
+	listOptions := entities.ImageListOptions{}
+
+	images, err := registry.ImageEngine().List(registry.GetContext(), listOptions)
+	if err != nil {
+		return nil, cobra.ShellCompDirectiveError
+	}
+
+	for _, image := range images {
+		for _, repo := range image.RepoTags {
+			if toComplete == "" {
+				//suggest only full repo path if no input is given
+				suggestions = append(suggestions, repo)
+			} else {
+				//suggested "registry.fedoraproject.org/f29/httpd:latest" as
+				// - "registry.fedoraproject.org/f29/httpd:latest"
+				// - "registry.fedoraproject.org/f29/httpd"
+				// - "f29/httpd:latest"
+				// - "f29/httpd"
+				// - "httpd:latest"
+				// - "httpd"
```

Comment:
I would strongly encourage you to add image IDs: not all images have names. (You should probably check for that)

Reply:
I think all image IDs should be listed, honestly: there are often cases where intermediate images have no tags, and someone running `podman rmi <TAB>` might want to see the full list of both IDs and tags. But this is just an opinion based on gut feel, and I don't have any evidence to support it. I think it would be OK to do what you think is best, and add a big FIXME comment saying that this needs UX testing. @mheon @rhatdan @TomSweeneyRedHat WDYT?

Reply:
I'm a little worried that printing all IDs could get overwhelming, it's easy to accumulate a lot of dangling images by using `podman build`. I would expect us to offer tab completion once a few characters are typed, though.